### PR TITLE
[RNMobile] Audio Player UI for audio block

### PR DESCRIPTION
**Pull request by etoledom**
`Gutenberg Mobile` https://github.com/wordpress-mobile/gutenberg-mobile/pull/3037
`WPAndroid` https://github.com/wordpress-mobile/WordPress-Android/pull/13843
`WPiOS` https://github.com/wordpress-mobile/WordPress-iOS/pull/15694

## Description

This PR adds a placeholder-ish Audio Player UI for the Audio Block.

## How has this been tested?

- Add an audio block.
- Check that the Audio Player UI looks as expected.

### Success
Light Mode | Dark Mode 
--------|-------
 <kbd></kbd>  |    <kbd></kbd>
   
### Uploading

Light Mode | Dark Mode 
--------|-------
<kbd></kbd>   |    <kbd></kbd>

### Error

Light Mode | Dark Mode 
--------|-------
 <kbd></kbd>   |    <kbd></kbd>

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by etoledom** [comment]
cc @iamthomasbishop @hypest 

To be clear, this has no functionality. The only "important" piece of functionality missing is the calculation of the audio file's time length, but we can obtain it from the `getMedia` request to the WordPress backend:

```json
            {
                "ID": 3865,
                "//": "...",
                "file": "christmas-time-6.mp3",
                "mime_type": "audio/mpeg",
                "extension": "mp3",
                "length": 238, <---
                "exif": {
                    "//": "...",
                    "length": 238, <---
                    "length_formatted": "3:58", <---
                },
                "//": "...",
            }
```

I would do this after this approach is accepted.

**Comment by iamthomasbishop** [comment]
Nice work @etoledom! I think this should work pretty nicely visually! My one concern is that there could be edge cases where this type of placeholder/façade UI will be confusing — such as when there's an error. I suppose this is pretty similar to how the Video block works so I don't expect any major issues, but let's make sure consider the various stress cases. For example:

- Error while uploading
- Progress/uploading
- Audio file is unavailable/unreachable (offline?)

For all of these examples, I think we can probably simply replace the "play" button depending on state:

- Error while uploading: use "error" icon
- Progress/uploading: use circular indicator @jd-alexander is working on or use RN [ActivityIndicator](https://reactnative.dev/docs/activityindicator)
- Unavailable/unreachable: use "error" icon

**Comment by hypest** [comment]
> this type of placeholder/façade UI will be confusing — such as when there's an error.

If I may add, even if not an error has occurred, I think it might be confusing anyway since the buttons won't be actionable at all. For example, even if they look disabled, it's not clear why they are disabled or how can someone enable them.

> I suppose this is pretty similar to how the Video block works so I don't expect any major issues

I agree that the video block is semi-fake as well, but user can indeed tap the "Play" button and playback will commence in a browser screen. On the Audio block, tapping the Play will do nothing at all, right?

**Comment by ceyhun** [comment]
Hey there 👋 I'll be continuing the work here.

> On the Audio block, tapping the Play will do nothing at all, right?

That's right and it's the same behavior on web. But do we want the mobile audio block to play audio in the editor the end? That could introduce some other complexities and challenges as mentioned in the PR description. @iamthomasbishop WDYT? Should we maybe make the current "fake" disabled UI look more like it cannot be interacted instead?

**Comment by jd-alexander** [comment]
Thanks for these thoughts @ceyhun 

> That's right and it's the same behavior on web. But do we want the mobile audio block to play audio in the editor the end? 

I honestly think this could become a possibility in the future. 
My thinking is that if we ever need to implement this, if say the web is doing it, we will have to find a way to address those issues. That seems like a mini-project on its own especially if we went the route of building our own audio player. 

>  WDYT? Should we maybe make the current "fake" disabled UI look more like it cannot be interacted instead?

Yes, it will need to look disabled so the user doesn't try to interact with it. Honestly, now when I think about it, I would prefer if the Audio block looked more like the File block ( with a few minor alterations like adding an Audio Icon or something to it). The reason, I am thinking this is we are attempting to create a disabled UI similar to the web but as @iamthomasbishop stated there are several cases where the player UI could become confusing. So going down the File block path would allow us to avoid these cases as we already solved for them already. Of course, this would deviate away from the experience of the web a bit but I think that's an okay compromise for now since the functionality being offered by both blocks on their respective platforms would be the same. The work done already could be utilized in a project that would require the Audio block to have playback functionality. Let me know what you think peeps.

**Comment by jd-alexander** [comment]
> > So going down the File block path would allow us to avoid these cases as we already solved for them already.
> 
> 
> 
> As I see it, the final goal of this it is to deliver the WISIWIG experience. It's not as much as it look similar to the Web editor, but closer to the final published post. This concept is mainly visual, so it's fine if the editor won't play audios on the editor.
> 
Ohh I never thought of it from this point of view. Thanks for this.  
> 
> 
> > WDYT? Should we maybe make the current "fake" disabled UI look more like it cannot be interacted instead?
> 
> 
> 
> I agree that this would be good to avoid some confusion 👍 
> 
I agree with this as well. 
> 
> 
> (Just wanted to share my thoughts of why I started this PR and why I thought it could be a good approach 😃 )

Appreciated @etoledom 😄

**Comment by ceyhun** [comment]
Thanks everyone! Then I guess it makes sense to continue with the current approach for now.

> Should we maybe make the current "fake" disabled UI look more like it cannot be interacted instead?

@iamthomasbishop any ideas about how to achieve this?

> For all of these examples, I think we can probably simply replace the "play" button depending on state:
> 
> Error while uploading: use "error" icon
> Progress/uploading: use circular indicator @jd-alexander is working on or use RN ActivityIndicator
> Unavailable/unreachable: use "error" icon

@iamthomasbishop also again I couldn't really picture this because compared to the Video block we have in the Audio block also the slider, time indicators and volume controls. Could you maybe share some visuals?

**Comment by jd-alexander** [comment]
@iamthomasbishop I wanted to provide some more context about some of the suggestions you made:

> Progress/uploading: use circular indicator @jd-alexander is working on or use RN ActivityIndicator

We won't be shipping that circular indicator this time around as there are some upload flow improvements that need to be made and optimizing a new control for the current functionality won't be beneficial for us long term. More context about this in Eduardo's comment [here](https://github.com/WordPress/gutenberg/pull/27429#issuecomment-741910197)

So with that said, would utilizing the current upload progress UI/UX work with replacing the play button in the disabled state of the control?

**Comment by iamthomasbishop** [comment]
@ceyhun 👋 Until we're ready to implement a true inline player UI, I think it might be prudent to design something explicitly "placeholder-ish", rather than try to replicate the native player. I spent a few minutes sketching out an approach that I think could work, which borrows elements both from the placeholder/starter state of various media blocks and also the general layout of our Cell component. Here's a quick sketch of what I'm imagining for the component itself (which could be re-used for other scenarios where we want to show a "compact" representation:

<a href="https://user-images.githubusercontent.com/1200432/101817759-7b46e700-3ae8-11eb-87ce-6d5dec39b372.jpeg"></a>

And based on that, here are a bunch of variations, using either icon or text button as the right accessory:

<a href="https://user-images.githubusercontent.com/1200432/101817909-bb0dce80-3ae8-11eb-9b28-d0973a25b24f.jpeg"></a>

I'm leaning towards a text label (middle column) because the primary action is very clear. In terms of the copy for that label, my suggestion would be either "Listen", "Open", or "View" — "Preview" might be confused with post/page preview, although a pre-pended/appended "play" icon on the text label might be enough of an affordance to avoid that issue.

Here are some quick higher-fidelity mocks to get a feel (note: colors, sizing, etc. not final):

<a href="https://user-images.githubusercontent.com/1200432/101821463-99631600-3aed-11eb-9721-461f81a9d65f.png"></a>

I'd love to hear y'all's thoughts and/or suggestions! // cc @ceyhun @etoledom @jd-alexander and @kyleaparker

**Comment by jd-alexander** [comment]
Thanks for these high fidelity mockups @iamthomasbishop I am in agreement with the placeholder UI being represented like this until we have an actual player. I am in favor of the rounded designs on the right because it looks similar to the preview of the Audio block on Chrome. I also like the text label as the primary action so we can be clear on what the user will be able to perform.

Would we still utilize the approach of changing the main icon based on the state?

**Comment by ceyhun** [comment]
> I am in favor of the rounded designs on the right because it looks similar to the preview of the Audio block on Chrome. I also like the text label as the primary action so we can be clear on what the user will be able to perform.

I agree with both of these. 

> Would we still utilize the approach of changing the main icon based on the state?

## Video Block
Uploading | Error | Success
-- | -- | --
||

## File Block
Uploading | Error | Success
-- | -- | --
||


Looking at the Video block, I see that it's closer to its success state and the option to retry is more visible. Audio block could be similar to File block's error state if we only change the icon. In the error state you can still press the File block again to get the retry options.

I think progress bar will come for free with Audio block as well so maybe we show the success UI for the uploading state with the progress bar at the top? I'm still not sure how to show the error icon and the message to indicate that it can be retried though.

**Comment by jd-alexander** [comment]
> I think progress bar will come for free with Audio block as well so maybe we show the success UI for the uploading state with the progress bar at the top?

I agree with using the current progress bar. So we might end up having to use the "Normal" design because the progress bar would flush beautifully with its container.

**Comment by iamthomasbishop** [comment]
@jd-alexander @ceyhun Apologies for the delay, finally got back to this today. I spent some time right before the break pulling together some sketches of the various states to wrap my head around this.

Here's (roughly) what I'm thinking for the various states:

<a href="https://user-images.githubusercontent.com/1200432/104075433-6d17e380-51d8-11eb-9851-ef8385d9afb7.jpeg"></a>

In terms of layout for this "minimal" version of the audio player UI, I was imagining something a bit closer to the File block — where everything is left-aligned and "horizontal". Here are a few variations on that theme:

<a href="https://user-images.githubusercontent.com/1200432/104075468-7ef98680-51d8-11eb-9efc-7bb98068194b.jpeg"></a>


In terms of the progress bar, I am leaning towards the one that has the progress bar across the top because I agree our goal should be to reuse the existing component. *Side note: If we go with this approach, it'd probably make more sense to not do the rounded corners.*

One difference here is the uploading state — for the File and Audio blocks, I think we should probably add a text label of "Uploading..." (or similar) that sits in the same location as the error message (below the title). This would result in a static block height, instead of jumping around between uploading, error, and success states.

Here are some more detailed sketches of each state of the Audio vs. File block. Let me know if you have questions or something isn't quite right.

<a href="https://user-images.githubusercontent.com/1200432/104075513-98023780-51d8-11eb-9631-9f2c593ee465.jpeg"></a>

**Comment by ceyhun** [comment]
Thanks @iamthomasbishop!

> In terms of the progress bar, I am leaning towards the one that has the progress bar across the top because I agree our goal should be to reuse the existing component. Side note: If we go with this approach, it'd probably make more sense to not do the rounded corners.

The one with the progress bar on top makes more sense to me too. I guess in the end it would be like this one you shared before then? 

audio-no-pill

**Comment by iamthomasbishop** [comment]
> I guess in the end it would be like this one you shared before then?

@ceyhun Yup, exactly! 😊 I was thinking the straight-cornered progress bar might look a little odd with slightly-rounded corners on the block background, but maybe that's a non-issue if the block boundary has `overflow: hidden`?

**Comment by ceyhun** [comment]
> I was thinking the straight-cornered progress bar might look a little odd with slightly-rounded corners on the block background, but maybe that's a non-issue if the block boundary has overflow: hidden? (Assuming the progress bar is inside the boundary)

There's another component that's rendering the progress bar and the UI underneath. Should be showing the straight-cornered progress bar by default. But I'll try to pass a custom style argument to it to override this case.

Another question @iamthomasbishop: The progress bar seems will be looking the same in dark mode for file? Do you think that would look fine for audio as well?

**Comment by iamthomasbishop** [comment]
> There's another component that's rendering the progress bar and the UI underneath

@ceyhun  Ah, okay. My concern was that if you have the progress bar and UI stacked, we'd get a little 4px "jump" when the progress bar disappears (to avoid this w/ the File block, I think we ended up translating the Y-position of the bar downwards so the top edges aligned). Here's a little example to illustrate.

<a href="https://user-images.githubusercontent.com/1200432/104661522-4bae7000-568e-11eb-8f41-a2139363877a.png"></a>

> Should be showing the straight-cornered progress bar by default. But I'll try to pass a custom style argument to it to override this case.

If we aren't able to easily add the progress bar inside the "wrapper", I would be okay with either sticking with straight corners or potentially even zeroing out the border-radii from the top-left and top-right corners of the player UI while the progress bar is visible — here's an example:

<a href="https://user-images.githubusercontent.com/1200432/104662034-84027e00-568f-11eb-99f4-533033b14c49.png"></a>

> Another question @iamthomasbishop: The progress bar seems will be looking the same in dark mode, similar to file. Do you think that would look fine for audio as well?

Just to clarify, are we using the same colors on other progress bars in dark mode, or is the File block unique? Assuming the File block is using the same colors as other progress bars (on video, image, etc.), let's just roll with the same styling as that for now. At some point it would be nice to refine that progress bar component to use dark-mode-friendly colors but let's just use what we have for now.

---

I still need to finish some of the detailed specs, which I'll drop in this ticket tomorrow morning. :) Thanks Ceyhun!

**Comment by ceyhun** [comment]
> @ceyhun Ah, okay. My concern was that if you have the progress bar and UI stacked, we'd get a little 4px "jump" when the progress bar disappears (to avoid this w/ the File block, I think we ended up translating the Y-position of the bar downwards so the top edges aligned). Here's a little example to illustrate.

Ah, that's right! I'll make sure it doesn't jump 👍 

> If we aren't able to easily add the progress bar inside the "wrapper", I would be okay with either sticking with straight corners or potentially even zeroing out the border-radii from the top-left and top-right corners of the player UI while the progress bar is visible — here's an example:

I think this would be easier, might go with this instead. Thanks!

> Just to clarify, are we using the same colors on other progress bars in dark mode, or is the File block unique?

It's the same on other progress bars. It's seems to be used in cover, file, gallery, image, video to be exact.

**Comment by ceyhun** [comment]
> I still need to finish some of the detailed specs, which I'll drop in this ticket tomorrow morning. :) Thanks Ceyhun!

Hey there @iamthomasbishop, I've implemented the functionality and the basic layout here, but I still need the detailed specs for color, fonts, sizes, etc.

**Comment by iamthomasbishop** [comment]
Hey @ceyhun & @jd-alexander ! I created [a page on the design docs](https://zeroheight.com/226850145/p/89904f-audio-block-wip) to house the specs for the Audio placeholder UI and dropped some notes there as well. Let me know if y'all have any questions!

**Comment by jd-alexander** [comment]
> Hey @ceyhun & @jd-alexander ! I created a page on the design docs to house the specs for the Audio placeholder UI and dropped some notes there as well. 

This design guide is spot on @iamthomasbishop . Thanks much!

> Let me know if y'all have any questions!

Sure 🙇‍♂️

**Comment by jd-alexander** [comment]
The PR description was updated with screenshots from the latest version of the design based on the specs and the supporting code has been added to the PR. 

All of the sizing and spacing requirements haven't been implemented in totality as yet. The progress bar's height needs to be moved from 6px to 4px.

**Comment by ceyhun** [comment]
> @ceyhun The PR description was updated with screenshots from the latest version of the design based on the specs and the supporting code has been added to the PR.

Thanks for the updates @jd-alexander!

> All of the sizing and spacing requirements haven't been implemented in totality as yet. The progress bar's height needs to be moved from 6px to 4px.

Fixed this and the top-left and top-right corners when uploading by overriding the `MediaUploadProgress` styles.

> The error container needs to be styled differently so that the vertical spacing is less. I wasn't sure how to accomplish that so some help with that would be greatly appreciated.

I used the same `Text` component for uploading, success, error captions and have put it in a container with an optional icon. Though I remember there was a reason for using `PlainText` for error messages instead of `Text`, but right now it seemed fine to me with `Text`. Also used a custom retry message without new lines for this.

**Comment by paaljoachim** [comment]
I might have missed a comment about this....
It looks like the mobile version of the Audio block is getting a nice UI upgrade.
Are there also plans on adding a somewhat similar version of the UI for the mobile Audio block to the desktop version? As it would be nice to have the work done here carry over to the desktop version as well.

Current Audio block for the desktop looks like this in the backend in Gutenberg version 9.8 with Twenty Twenty One.
Screen Shot 2021-01-23 at 14 01 04

Turning the above into something like this:
Audio-block

**Comment by iamthomasbishop** [comment]
@paaljoachim Good question! This is a temporary static/“placeholder” UI, sans playback control. At the moment, we’re relying on the system media picker to do the heavy lifting — if you tap the placeholder’s action button, it launches the system media player. I believe the long-term plan is to display a standard system player inline and allow inline playback (just like the web editor).

Regarding a potential custom inline player UI, I’m hesitant to diverge unless we have a good reason to — such as when a theme has a custom player UI. It could be beneficial to have support for this, but because we don’t have “full” CSS support on ReactNative, that could leave us in an awkward position with only partial style support.

**Comment by paaljoachim** [comment]
Hi @iamthomasbishop Thomas

> Regarding a potential custom inline player UI, I’m hesitant to diverge unless we have a good reason to — such as when a theme has a custom player UI. It could be beneficial to have support for this, but because we don’t have “full” CSS support on ReactNative, that could leave us in an awkward position with only partial style support.

Good point. I believe down the line that @ItsJonQ Jon will likely look at the default ReactNative players to see how we can natively improve the player. 

Thanks.

File: packages/components/src/mobile/audio-player/index.native.js [hunk 1]
```diff
@@ -119,7 +118,9 @@ function Player( {
 	let title = '';
 	let extension = '';
 
-	if ( fileName ) {
+	if ( src ) {
+		const decodedURI = decodeURI( src );
```
Comment (jd-alexander) [inline packages/components/src/mobile/audio-player/index.native.js:122]: @ceyhun where does `decodeURI ` come from exactly?

Comment (jd-alexander) [reply packages/components/src/mobile/audio-player/index.native.js:122]: I also just saw a safe version of decodeURI https://github.com/WordPress/gutenberg/blob/004a32ec09d5b789be694d85725fe49cca385f08/packages/url/src/safe-decode-uri.js can that be helpful here in anyway?

Comment (jd-alexander) [reply packages/components/src/mobile/audio-player/index.native.js:122]: Oh @ceyhun `decodeURI` is a built-in Javascript function :) I just realized that.

Comment (ceyhun) [reply packages/components/src/mobile/audio-player/index.native.js:122]: It's built-in in JS, you can check the reference [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects#function_properties) for some others. BTW I'm not really happy parsing the file name and extension from the media src here, I would appreciated any other ideas or suggestions 🙏

Comment (ceyhun) [reply packages/components/src/mobile/audio-player/index.native.js:122]: >I also just saw a safe version of decodeURI 004a32e/packages/url/src/safe-decode-uri.js can that be helpful here in anyway?

Good catch! Should better use this one 👍

Comment (jd-alexander) [reply packages/components/src/mobile/audio-player/index.native.js:122]: > It's built-in in JS, you can check the reference here for some others.

okay understood. Thanks for the reference.

> . BTW I'm not really happy parsing the file name and extension from the media src here, I would appreciated any other ideas or suggestions 🙏

Could you expound on this a bit? Do you mean that you aren't happy with the location it is in? Or you are not sure if the output/result of this parsing will be reliable and user friendly? 

Another question I wanted to ask is could we check to see if the media title is present and we use that as the title and if it's not present we use the filename that's part of the URI. Let me know if this helps in anyway. I am looking at the media object and the overall logic to see if I can find any more clues to assist.

Comment (jd-alexander) [reply packages/components/src/mobile/audio-player/index.native.js]: I am probably misinterpreting this @ceyhun is it that when the file name and extension are being parsed sometimes it's missing which is leading to the title of the audio file being used as the extension?

File: packages/components/src/mobile/audio-player/styles.scss [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+	border-radius: 500px;
+	background-color: $light-ultra-dim;
+}
+
+.buttonBackgroundDark {
+	background-color: $dark-ultra-dim;
+}
+
+.buttonText {
+	font-size: 17;
```
Comment (jd-alexander) [inline packages/components/src/mobile/audio-player/styles.scss:99]: We should make this font-size 14 based on a discussion Thomas had about Android styling. how is it looking on iOS? This is how it looks on Android.

File: packages/components/src/mobile/audio-player/index.native.js (hunk 1 above)
Comment (ceyhun) [reply packages/components/src/mobile/audio-player/index.native.js]: > Or you are not sure if the output/result of this parsing will be reliable and user friendly?

Closer to this ☝️ 

> Another question I wanted to ask is could we check to see if the media title is present and we use that as the title and if it's not present we use the filename that's part of the URI. Let me know if this helps in anyway. I am looking at the media object and the overall logic to see if I can find any more clues to assist.
 
Was using the media title before, but it was not there when selecting from media library.

> I am probably misinterpreting this @ceyhun is it that when the file name and extension are being parsed sometimes it's missing the extension which is leading to the title of the audio file being used as the extension?

If we use `src` which is the URI of the file when picking from the phone's local storage and URL of the file picking from media library. In both cases for iOS and Android `src` had a filename and extension in the URI+URL. But the filename isn't always the same as the file title assigned to the file when it's uploaded to the remote from local storage.

File: packages/components/src/mobile/audio-player/styles.scss (hunk 2 above)
Comment (ceyhun) [reply packages/components/src/mobile/audio-player/styles.scss:99]: Nice catch! I think I missed that as well. Here's how it looks on iOS: 

Will commit the sugesstion 👍

**Review by jd-alexander — APPROVED** [review]
The behavior here was tested and the changes work as expected. LGTM 🚢

**Pull request merged**